pkg/virt-controller/services/rendervolumes.go

```diff
@@ -332,57 +332,95 @@ func withAccessCredentials(accessCredentials []v1.AccessCredential) VolumeRender
 	}
 }
 
-func withTPM(vmi *v1.VirtualMachineInstance) VolumeRendererOption {
+func PathForSwtpm(vmi *v1.VirtualMachineInstance) string {
+	swtpmPath := "/var/lib/libvirt/swtpm"
+	if util.IsNonRootVMI(vmi) {
+		swtpmPath = filepath.Join(util.VirtPrivateDir, "libvirt", "qemu", "swtpm")
+	}
+
+	return swtpmPath
+}
+
+func PathForSwtpmLocalca(vmi *v1.VirtualMachineInstance) string {
+	localCaPath := "/var/lib/swtpm-localca"
+	if util.IsNonRootVMI(vmi) {
+		localCaPath = filepath.Join(util.VirtPrivateDir, "var", "lib", "swtpm-localca")
+	}
+
+	return localCaPath
+}
+
+func PathForNVram(vmi *v1.VirtualMachineInstance) string {
+	nvramPath := "/var/lib/libvirt/swtpm"
```

Comment:
Hey, I was just wondering if it would cause errors for the virt-launcher Pod when both TPM and NVRAM persistence are enabled. For example, in this case the volumeMounts of the virt-launcher Pod would be like:

```
    - mountPath: /var/lib/libvirt/swtpm
      name: vm1-state
      subPath: swtpm
   - mountPath: /var/lib/libvirt/swtpm
      name: vm1-state
      subPath: nvram
```

Might this be rejected by kubernetes?

Reply:
Great question, and yes it's ok to mount a given volume in multiple places, especially when using different sub-paths every time.
In fact, TPM alone mounts it twice. Please note that there is a functional test for exactly this (persistent TPM+EFI).
Thanks for the review!

Reply:
I think the problem would be that it's mounting the volume to the same path "/var/lib/libvirt/swtpm" for TPM and NVRAM.

```
volumeMounts:
- mountPath: /var/lib/libvirt/swtpm <------------------ this is the same with the one for NVRAM
  name: vm1-state
  subPath: swtpm
- mountPath: /var/lib/libvirt/swtpm <------------------ 
  name: vm1-state
  subPath: nvram
```

This would cause the virt-launcher Pod to be rejected.

For example, the below Pod will be rejected with the error `The Pod "p1" is invalid: spec.containers[0].volumeMounts[1].mountPath: Invalid value: "/usr/share/abc": must be unique`:

```
apiVersion: v1
kind: Pod
metadata:
  name: p1
spec:
  containers:
  - image: ubuntu
    command:
      - "sleep"
      - "604800"
    name: ubuntu
    volumeMounts:
    - mountPath: "/usr/share/abc"
      name: v1
      subPath: sp1
    - mountPath: "/usr/share/abc"
      name: v1
      subPath: sp2
  volumes:
  - name: v1
    persistentVolumeClaim:
      claimName: pvc-1
```

I guess the reason the functional test is passing is because it's running in non-root mode, in which case the `volumeMounts` of the virt-launcher Pod is something like

```
    - mountPath: /var/run/kubevirt-private/libvirt/qemu/swtpm
      name: testvmi-7j99w-state
      subPath: swtpm
    - mountPath: /var/run/kubevirt-private/libvirt/qemu/nvram
      name: testvmi-7j99w-state
      subPath: nvram
```

Would it make sense to change `nvramPath := "/var/lib/libvirt/swtpm"` to `nvramPath := "/var/lib/libvirt/nvram"` in the code? Or was this just a typo?

Reply:
You're absolutely right, awful typo there :facepalm: 
Great catch, and sorry for not reading your initial comment thoroughly enough.
Will fix asap...

Reply:
Btw, I think we might need to check if this path exists or not in the base image -- `/var/lib/libvirt/nvram`. If it does not exist, I guess libvirt may not create the directory for us and thus would fail...

Reply:
Nice catch, again... The right path, that will get created by libvirt if it doesn't exist, is `/var/lib/libvirt/qemu/nvram` and not `/var/lib/libvirt/nvram` (swtpm is `/var/lib/libvirt/swtpm`, which threw me off). Fixed.